<a href="https://colab.research.google.com/github/iambrookedrake/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/Brooke_Drake_DSPT6_U1S2M1_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

In [0]:
### YOUR CODE STARTS HERE
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/ --no-check-certificate


--2020-04-14 02:51:03--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 653 [text/html]
Saving to: ‘index.html’

index.html          100%[===================>]     653  --.-KB/s    in 0s      

2020-04-14 02:51:03 (23.2 MB/s) - ‘index.html’ saved [653/653]



In [0]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data --no-check-certificate

--2020-04-14 02:51:11--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.1s    

2020-04-14 02:51:12 (137 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [0]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind
import matplotlib.pyplot as plt
import seaborn as sns

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [0]:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.names --no-check-certificate

--2020-04-14 02:51:16--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.names
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 6868 (6.7K) [application/x-httpd-php]
Saving to: ‘house-votes-84.names’

house-votes-84.name 100%[===================>]   6.71K  --.-KB/s    in 0s      

2020-04-14 02:51:16 (117 MB/s) - ‘house-votes-84.names’ saved [6868/6868]



In [0]:
columns = ['party', 'handicapped-infants', 'water-project-cost-sharing', 
           'adoption-of-the-budget-resolution', 'physician-fee-freeze', 
           'el-salvador-aid', 'religious-groups-in-schools', 
           'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile',
           'immigration', 'synfuels-corporation-cutback', 'education-spending', 
           'superfund-right-to-sue', 'crime', 'duty-free-exports', 
           'export-administration-act-south-africa']

In [0]:
votes = pd.read_csv('house-votes-84.data', header=None, names=columns)
votes = votes.replace({'y':1, 'n':0})
votes = votes.replace('?', np.nan)
print(votes.shape)
votes.head()

(435, 17)


party  ...  export-administration-act-south-africa
0  republican  ...                                     1.0
1  republican  ...                                     NaN
2    democrat  ...                                     0.0
3    democrat  ...                                     1.0
4    democrat  ...                                     1.0

[5 rows x 17 columns]

In [0]:
rep = votes[votes['party']=='republican']
print(rep.shape)
rep.head()

(168, 17)


party  ...  export-administration-act-south-africa
0   republican  ...                                     1.0
1   republican  ...                                     NaN
7   republican  ...                                     1.0
8   republican  ...                                     1.0
10  republican  ...                                     0.0

[5 rows x 17 columns]

In [0]:
dem = votes[votes['party']=='democrat']
print(dem.shape)
dem.head()

(267, 17)


party  ...  export-administration-act-south-africa
2  democrat  ...                                     0.0
3  democrat  ...                                     1.0
4  democrat  ...                                     1.0
5  democrat  ...                                     1.0
6  democrat  ...                                     1.0

[5 rows x 17 columns]

In [0]:
def ttestcol(col):
   t,p= ttest_ind(dem[dem.columns[col]], rep[rep.columns[col]], nan_policy='omit')
   column=votes.columns[col]
   print(column, "issue has TStatistic of",t, "and the Pvalue is",p)

In [0]:
ttestcol(3)

adoption-of-the-budget-resolution issue has TStatistic of 23.21277691701378 and the Pvalue is 2.0703402795404463e-77


In [0]:
for i in range(1,dem.shape[1]): 
    ttestcol(i)

handicapped-infants issue has TStatistic of 9.205264294809222 and the Pvalue is 1.613440327937243e-18
water-project-cost-sharing issue has TStatistic of -0.08896538137868286 and the Pvalue is 0.9291556823993485
adoption-of-the-budget-resolution issue has TStatistic of 23.21277691701378 and the Pvalue is 2.0703402795404463e-77
physician-fee-freeze issue has TStatistic of -49.36708157301406 and the Pvalue is 1.994262314074344e-177
el-salvador-aid issue has TStatistic of -21.13669261173219 and the Pvalue is 5.600520111729011e-68
religious-groups-in-schools issue has TStatistic of -9.737575825219457 and the Pvalue is 2.3936722520597287e-20
anti-satellite-test-ban issue has TStatistic of 12.526187929077842 and the Pvalue is 8.521033017443867e-31
aid-to-nicaraguan-contras issue has TStatistic of 18.052093200819733 and the Pvalue is 2.82471841372357e-54
mx-missile issue has TStatistic of 16.437503268542994 and the Pvalue is 5.03079265310811e-47
immigration issue has TStatistic of -1.735911732

In [0]:
dem.describe()

handicapped-infants  ...  export-administration-act-south-africa
count           258.000000  ...                              185.000000
mean              0.604651  ...                                0.935135
std               0.489876  ...                                0.246956
min               0.000000  ...                                0.000000
25%               0.000000  ...                                1.000000
50%               1.000000  ...                                1.000000
75%               1.000000  ...                                1.000000
max               1.000000  ...                                1.000000

[8 rows x 16 columns]

In [0]:
dem.columns

Index(['party', 'handicapped-infants', 'water-project-cost-sharing',
       'adoption-of-the-budget-resolution', 'physician-fee-freeze',
       'el-salvador-aid', 'religious-groups-in-schools',
       'anti-satellite-test-ban', 'aid-to-nicaraguan-contras', 'mx-missile',
       'immigration', 'synfuels-corporation-cutback', 'education-spending',
       'superfund-right-to-sue', 'crime', 'duty-free-exports',
       'export-administration-act-south-africa'],
      dtype='object')

#2
Based off the t-stat of -49.36 and p-value of 1.99e-177, 'physician-fee-freeze' was more favored by Democrats than Republicans.

#3
Based off the t-stat of 23.2 and p-value of 2.07e-77, it seems 'adoption-of-the-budget-resolution' was an issue Republicans favored more than Democrats

#4 
No Issue has true bipartisan support with 95% confidence. The closest example would be export-administration-act-south-africa with a t-stat of 6.849454815841208 and a Pvalue of 3.65e-11


## Stretch Goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Then check your Answers using Scipy!